src/examples/4626-router/Generic4626Router.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {
+    toBeforeSwapDelta,
+    BeforeSwapDelta,
+    BeforeSwapDeltaLibrary
+} from "v4-core/src/types/BeforeSwapDelta.sol";
+import {Currency} from "v4-core/src/types/Currency.sol";
+import {PoolKey} from "v4-core/src/types/PoolKey.sol";
+import {PoolId} from "v4-core/src/types/PoolId.sol";
+import {SafeCast} from "v4-core/src/libraries/SafeCast.sol";
+import {IHooks} from "v4-core/src/interfaces/IHooks.sol";
+import {Hooks} from "v4-core/src/libraries/Hooks.sol";
+import {IPoolManager} from "v4-core/src/interfaces/IPoolManager.sol";
+
+import {BaseHook} from "uniswap-hooks/base/BaseHook.sol";
+import {CurrencySettler} from "uniswap-hooks/utils/CurrencySettler.sol";
+
+import {DeltaResolver} from "v4-periphery/src/base/DeltaResolver.sol";
+
+import {ERC4626, ERC20} from "solmate/src/mixins/ERC4626.sol";
+
+/// @title Generic Router for ERC4626 Token Wrappers
+/// @dev Only supports symmetric ERC4626 Vaults
+contract Generic4626Router is BaseHook {
+    using SafeCast for int256;
+    using SafeCast for uint256;
+
+    error Generic4626Router__NotAllowed();
+    error Generic4626Router__InvalidPoolFee();
+
+    struct PoolDetails {
+        bool isInitialized;
+        bool wrapDirection; // true if zeroToOne, false if oneToZero
+        ERC4626 vault;
+        ERC20 underlying;
```

Comment:
have a hunch that this doesnt need to be storage, if you have the pool key and know the wrap direction, you can always deduce if `key.currency0` is the vault (or not)

Reply:
its possible to write an abstraction library:

```solidity
(Currency vault, Currency underlying) = poolKey.getVaultUnderlying(details.wrapDirection);
```

Reply:
This is correct, that's how I wrote this initially. (See comment in the if/else scope)

I do think it's preferable to be optimized here, so will switch it back to that.